packages/job-worker/src/ingest/bucketAdlibs.ts

```diff
@@ -134,155 +153,161 @@ export async function handleBucketItemImport(context: JobContext, data: BucketIt
 	let newRank: number | undefined = undefined
 	let onlyGenerateOneItem = false
 
-	const ps: Promise<any>[] = []
-	let isFirstShowStyleVariant = true
-	for (const showStyleVariant of showStyleVariants) {
-		const showStyleCompound = createShowStyleCompound(showStyleBase, showStyleVariant)
-
-		if (!showStyleCompound)
-			throw new Error(`Unable to create a ShowStyleCompound for ${showStyleBase._id}, ${showStyleVariant._id} `)
-
-		const watchedPackages = WatchedPackagesHelper.empty(context)
-
-		const contextForVariant = new ShowStyleUserContext(
-			{
-				name: `Bucket Ad-Lib`,
-				identifier: `studioId=${context.studioId},showStyleBaseId=${showStyleBase._id},showStyleVariantId=${showStyleVariant._id}`,
-				tempSendUserNotesIntoBlackHole: true, // TODO-CONTEXT
-			},
-			context,
-			showStyleCompound,
-			watchedPackages
-		)
+	await context.directCollections.runInTransaction(async (transaction) => {
+		const ps: Promise<any>[] = []
+		let isFirstShowStyleVariant = true
+		for (const showStyleVariant of showStyleVariants) {
+			const showStyleCompound = createShowStyleCompound(showStyleBase, showStyleVariant)
```

Comment:
This loop has now become very deeply nested. Maybe this could be extracted to it's own function that deals with an individual `showStyleVariant`, for clarity?

Reply:
I dont think it is that simple to refactor this to achieve that. 
Each run of the loop modifies the `newRank` variable and adds to the `ps` array. The first run will set `onlyGenerateOneItem` and `isFirstShowStyleVariant`. And each run will modify one of `adlibIdsToRemove` or `actionIdsToRemove` depending what it was.
So to achieve that will take either a lot of parameters that may be mutated, or some complex return types that will need to be broken out and processed. Perhaps this is a good candidate for using a class, to allow for this shared state?

I have moved this method into its own file and broken out a couple of pretty isolated methods, and some other tidying. Going much further will mean largely rewriting it.